### PR TITLE
Change best effort queue behaviour in `dispute-coordinator`

### PR DESCRIPTION
**Pull request by tdimitrov**
Changes the implementation of best-effort queue in `dispute-coordinator`. Orders the disputes by the height of the relay chain parent (the same way priority queue works).

Fixes https://github.com/paritytech/polkadot/issues/5896

### CODE REVIEW

File: node/core/dispute-coordinator/src/participation/queues/mod.rs [hunk 1]
```diff
@@ -277,8 +247,9 @@ struct CandidateComparator {
 	/// just using the lowest `BlockNumber` of all available including blocks - the problem is,
 	/// that is not stable. If a new fork appears after the fact, we would start ordering the same
 	/// candidate differently, which would result in the same candidate getting queued twice.
-	relay_parent_block_number: BlockNumber,
+	relay_parent_block_number: Option<BlockNumber>,
 	/// By adding the `CandidateHash`, we can guarantee a unique ordering across candidates.
+	/// Additionally if `BlockNumber` can't be obtained the `CandidateHash` is used for ordering.
```
Comment (BradleyOlson64) [inline node/core/dispute-coordinator/src/participation/queues/mod.rs:252]: I think we also use CandidateHash for ordering if BlockNumber is the same between two disputes.

File: node/core/dispute-coordinator/src/participation/queues/mod.rs [hunk 2]
```diff
@@ -333,11 +306,28 @@ impl PartialOrd for CandidateComparator {
...
 	fn cmp(&self, other: &Self) -> Ordering {
-		match self.relay_parent_block_number.cmp(&other.relay_parent_block_number) {
-			Ordering::Equal => (),
-			o => return o,
+		return match (self.relay_parent_block_number, other.relay_parent_block_number) {
+			(None, None) => {
+				// No relay parents for both -> compare hashes
+				self.candidate_hash.cmp(&other.candidate_hash)
+			},
+			(Some(_), Some(_)) => {
```
Comment (BradleyOlson64) [inline node/core/dispute-coordinator/src/participation/queues/mod.rs:314]: Here's a refactor that would look more natural to me:

```
(Some(self_parent_number), Some(other_parent_number)) => {
    match self_parent_number.cmp(other_parent_number) => {
        // These lines same
    }
}
```

I can see why you might want to stick with the nice look of `(Some(_), Some(_))` though

**Review by BradleyOlson64 — APPROVED** [review]

File: node/core/dispute-coordinator/src/participation/queues/mod.rs (hunk 2 above)
Comment (tdimitrov) [reply node/core/dispute-coordinator/src/participation/queues/mod.rs:314]: Your suggestion looks better, thanks!

File: node/core/dispute-coordinator/src/participation/queues/mod.rs [hunk 3]
```diff
@@ -143,14 +129,13 @@ impl ParticipationRequest {
...
-		Self { best_effort: HashMap::new(), priority: BTreeMap::new() }
+		Self { best_effort: BTreeMap::new(), priority: BTreeMap::new() }
 	}
 
 	/// Will put message in queue, either priority or best effort depending on priority.
 	///
 	/// If the message was already previously present on best effort, it will be moved to priority
-	/// if it considered priority now, otherwise the `added_count` on the best effort queue will be
-	/// bumped.
+	/// if it considered priority now.
```
[the remark this answers is not in the document]
Comment (tdimitrov) [reply node/core/dispute-coordinator/src/participation/queues/mod.rs:138]: 'already previously' also sounds weird, I think :)

[the remark this answers is not in the document]
Comment (tdimitrov) [reply node/core/dispute-coordinator/src/participation/queues/mod.rs:138]: Yes, I missed 'is' here. Thanks.

File: node/core/dispute-coordinator/src/participation/queues/mod.rs (hunk 1 above)
Comment (tdimitrov) [reply node/core/dispute-coordinator/src/participation/queues/mod.rs:252]: The whole sentence is misplaced. I'll rework this paragraph.

File: node/core/dispute-coordinator/src/participation/queues/mod.rs [hunk 4]
```diff
@@ -159,94 +144,76 @@ impl Queues {
...
 			// Remove any best effort entry:
-			self.best_effort.remove(&req.candidate_hash);
+			self.best_effort.remove(&comparator);
 			self.priority.insert(comparator, req);
 		} else {
 			if self.best_effort.len() >= BEST_EFFORT_QUEUE_SIZE {
 				return Err(QueueError::BestEffortFull)
 			}
 			// Note: The request might have been added to priority in a previous call already, we
 			// take care of that case in `dequeue` (more efficient).
```
Comment (eskimor) [inline node/core/dispute-coordinator/src/participation/queues/mod.rs:183]: This is actually no longer true. Since we are using a `CandidateComparator` for both queues now, we can refrain from putting a participation request in the best effort queue if already present in the priority queue efficiently and don't need to resort to handling that in `dequeue`.

File: node/core/dispute-coordinator/src/participation/queues/mod.rs [hunk 5]
```diff
@@ -333,11 +307,28 @@ impl PartialOrd for CandidateComparator {
...
+				// No relay parents for both -> compare hashes
+				self.candidate_hash.cmp(&other.candidate_hash)
+			},
+			(Some(self_relay_parent_block_num), Some(other_relay_parent_block_num)) => {
+				match self_relay_parent_block_num.cmp(&other_relay_parent_block_num) {
+					Ordering::Equal => (),
+					o => return o,
+				}
+				// if the relay parent is the same for both -> compare hashes
+				self.candidate_hash.cmp(&other.candidate_hash)
```
Comment (eskimor) [inline node/core/dispute-coordinator/src/participation/queues/mod.rs:321]: Given that we already have a nested match, I think this would be easier to understand if that was in the `Ordering::Equal` match arm already.

File: roadmap/implementers-guide/src/node/disputes/dispute-coordinator.md [hunk 6]
```diff
@@ -359,14 +359,19 @@ times instead of just once to the oldest offender. This is obviously a good
...
-available we put participation on a best-effort queue, which at the moment is
-processed on the basis how often we requested participation locally, which
-equals the number of times we imported votes for that dispute. The idea is, if
-we have not seen the candidate included, but the dispute is valid, other nodes
-will have seen it included - so the more votes there are, the more likely it is
-a valid dispute and we should implicitly arrive at a similar ordering as the
-nodes that are able to sort based on the relay parent block height.
+For candidates we have not seen included, but we know are backed (thanks to chain
+scraping) or we have seen a dispute with 1/3+1 participation (confirmed dispute)
+on them - we put participation on a best-effort queue. It is has got the same
```
Comment (eskimor) [inline roadmap/implementers-guide/src/node/disputes/dispute-coordinator.md:364]: is has .. typo.

**Review by eskimor — COMMENTED** [review]
Nice!

File: node/core/dispute-coordinator/src/participation/queues/mod.rs (hunk 4 above)
Comment (tdimitrov) [reply node/core/dispute-coordinator/src/participation/queues/mod.rs:183]: Is this comment related to the `remove` in `dequeue` only? On line 159: https://github.com/paritytech/polkadot/blob/ab790ff09e3b73a93fae2559b9694ae4f66e7e15/node/core/dispute-coordinator/src/participation/queues/mod.rs#L159

Comment (tdimitrov) [reply node/core/dispute-coordinator/src/participation/queues/mod.rs:180]: Done.

Comment (eskimor) [reply node/core/dispute-coordinator/src/participation/queues/mod.rs:180]: Ok, I think there is a misunderstanding. Here in this else we should check whether the participation is already queued in priority and if so, skip enqueuing in the best effort queue.

Comment (eskimor) [reply node/core/dispute-coordinator/src/participation/queues/mod.rs:180]: It is the dual to the removal from the best effort above:

1. If it is already in best effort - remove it there if now enqueued in priority.
2. If it is already in priority, skip enqueuing in best-effort. 

How can this even happen? In case a participation request stays in the queue for a bit, dispute gets confirmed and finality kicks in - then we might queue again, but this time as best-effort instead of priority.

There is an argument that at this point it should not be priority anymore, but in any case we should only have it enqueued on one of the two queues. We might want to change 2 to:

2.  If it is already in priority - remove it there if now enqueued in best-effort.

So logic would be the same as for priority. It is a corner case and should not matter much in practice, so I don't mind either way.

Comment (tdimitrov) [reply node/core/dispute-coordinator/src/participation/queues/mod.rs:180]: Ohh right. There even is a test for it.

Comment (tdimitrov) [reply node/core/dispute-coordinator/src/participation/queues/mod.rs]: Should be fine now.

**Review by eskimor — APPROVED** [review]

File: node/core/dispute-coordinator/src/participation/queues/mod.rs [hunk 7]
```diff
@@ -159,94 +144,76 @@ impl Queues {
...
 			if self.best_effort.len() >= BEST_EFFORT_QUEUE_SIZE {
 				return Err(QueueError::BestEffortFull)
 			}
-			// Note: The request might have been added to priority in a previous call already, we
-			// take care of that case in `dequeue` (more efficient).
-			self.best_effort
-				.entry(req.candidate_hash)
-				.or_insert(BestEffortEntry { req, added_count: 0 })
-				.added_count += 1;
+			self.best_effort.entry(comparator).or_insert(req);
```
Comment (eskimor) [inline node/core/dispute-coordinator/src/participation/queues/mod.rs:184]: Actually a plain insert like in the priority case would suffice.

Comment (eskimor) [reply node/core/dispute-coordinator/src/participation/queues/mod.rs:184]: Can be fixed in the course of other work, no reason not to merge this one.

Comment (eskimor) [reply node/core/dispute-coordinator/src/participation/queues/mod.rs:184]: But what about that failing zombienet test? Zombienet problem or actual issue?

Comment (tdimitrov) [reply node/core/dispute-coordinator/src/participation/queues/mod.rs:184]: It's a small one. I'll fix it now.

The zombienet test fails because a metric can't be fetched:
```
2022-11-17T08:20:31.893Z zombie::network-node Connected to honest-validator-0
	 Error:  
		 Timeout(180), "getting desired parachain block height 2 within 180 secs".
    1) honest-validator-0: parachain 2000 block height is at least 2 within 180 seconds
2022-11-17T08:23:31.427Z zombie::network-node Connecting api for honest-validator-1 at ws://10.27.63.7:9944...
```
Looks like a connectivity issue to me. I don't see anything suspicious in the logs. There is a dispute opened and it concludes invalid.

Let's see if it will pass on the next run.

**Comment by tdimitrov** [comment]
bot merge
